src/charm.py

```diff
@@ -182,6 +189,45 @@ def _on_config_changed(self, _: ops.HookEvent, charm_state: CharmState) -> None:
         self.change_config(charm_state)
         self._set_workload_version()
 
+    @inject_charm_state
+    def _on_relation_joined(self, _: ops.HookEvent, charm_state: CharmState) -> None:
+        """Handle Synapse peer relation joined event.
+
+        Args:
+            charm_state: The charm state.
+        """
+        if charm_state.redis_config is None:
+            logger.debug("Scaling is not allowed without Redis integration.")
+            self.model.unit.status = ops.BlockedStatus(
+                "Redis integration not found. Please, verify it."
+            )
+            return
+        self.model.unit.status = ops.ActiveStatus()
+
+    @inject_charm_state
+    def _on_relation_departed(self, event: RelationDepartedEvent, charm_state: CharmState) -> None:
+        """Handle Synapse peer relation departed event.
+
+        Args:
+            event: relation departed event.
+            charm_state: The charm state.
+        """
+        if not self.unit.is_leader() or event.departing_unit == self.unit:
+            # returning because only the leader will change the charm status.
+            return
+        peer_relation = self.model.get_relation(synapse.SYNAPSE_PEER_RELATION_NAME)
+        if peer_relation is not None:
+            # if more than 1, check for Redis integration
+            synapse_units = len(peer_relation.units)
+            if synapse_units > 1:
+                logger.debug("Found %d units, checking for Redis integration.", synapse_units)
+                if charm_state.redis_config is None:
+                    logger.debug("Scaling is not allowed without Redis integration.")
+                    self.model.unit.status = ops.BlockedStatus(
+                        "Redis integration not found. Please, verify it."
+                    )
```

Comment:
I did not see it before, I think you missed a return here.

Reply:
Thanks!